File: VB/MsaglHelpers/MsaglGeometryGraphHelpers.vb
Imports System.Linq
Imports DevExpress.Diagram.Core
Imports DevExpress.Mvvm.Native
Imports System.Collections.Generic
Imports DevExpress.Diagram.Core.Layout
Imports Microsoft.Msagl.Core.Layout
Imports Microsoft.Msagl.Prototype.Phylo
Imports Microsoft.Msagl.Core.Geometry.Curves
Imports Microsoft.Msagl.Core.Geometry

Namespace MsaglHelpers
    Public NotInheritable Class MsaglGeometryGraphHelpers

        Private Sub New()
        End Sub

        Public Shared Function GetGetNodesPositionInfo(ByVal geometryGraph As GeometryGraph) As IEnumerable(Of PositionInfo(Of IDiagramItem))
            Return geometryGraph.Nodes.Select(Function(node) New PositionInfo(Of IDiagramItem)(DirectCast(node.UserData, IDiagramItem), Converter.MsaglPointToPointConvert(node.BoundingBox.LeftTop)))
        End Function
        Public Shared Function CreatePhyloTrees(ByVal graph As Graph(Of IDiagramItem)) As PhyloTree
            Dim phyloTree As New PhyloTree()
            For Each node In graph.Nodes
                AddNode(phyloTree, node)
            Next node
            For Each edge In graph.Edges
                AddPhyloTreeEdge(phyloTree, edge.From, edge.To, edge.Weight)
            Next edge
            Return phyloTree
        End Function
        Public Shared Function CreateGeometryGraph(ByVal graph As Graph(Of IDiagramItem)) As GeometryGraph
            Dim geomGraph As New GeometryGraph()
            For Each node In graph.Nodes
                AddNode(geomGraph, node)
            Next node
            For Each edge In graph.Edges
                AddEdge(geomGraph, edge.From, edge.To, edge.Weight)
            Next edge
            Return geomGraph
        End Function
        Public Shared Function AddNode(ByVal geometryGraph As GeometryGraph, ByVal item As IDiagramItem) As Node
            Dim msaglNode As Node = geometryGraph.FindNodeByUserData(item)
            If msaglNode Is Nothing Then
                msaglNode = New Node(CreateCurve(item), item)
                geometryGraph.Nodes.Add(msaglNode)
            End If
            Return msaglNode
        End Function
        Public Shared Sub AddEdge(ByVal geometryGraph As GeometryGraph, ByVal parentNodeSource As IDiagramItem, ByVal childNodeSource As IDiagramItem, ByVal weight As Double)
            geometryGraph.Edges.Add(New Edge(AddNode(geometryGraph, parentNodeSource), AddNode(geometryGraph, childNodeSource)) With {.Weight = CInt((weight))})
        End Sub
        Public Shared Sub AddPhyloTreeEdge(ByVal phyloTree As PhyloTree, ByVal parentNodeSource As IDiagramItem, ByVal childNodeSource As IDiagramItem, Optional ByVal weight As Double = 1)
            phyloTree.Edges.Add(New PhyloEdge(AddNode(phyloTree, parentNodeSource), AddNode(phyloTree, childNodeSource)) With {.Weight = CInt((weight))})
        End Sub
        Public Shared Function CreateCurve(ByVal item As IDiagramItem) As ICurve
            Return CurveFactory.CreateRectangle(item.ActualSize.Width, item.ActualSize.Height, New Point())
        End Function
    End Class
End Namespace

File: CS/MsaglHelpers/MsaglGeometryGraphHelpers.cs
using System.Linq;
using DevExpress.Diagram.Core;
using DevExpress.Mvvm.Native;
using System.Collections.Generic;
using DevExpress.Diagram.Core.Layout;
using Microsoft.Msagl.Core.Layout;
using Microsoft.Msagl.Prototype.Phylo;
using Microsoft.Msagl.Core.Geometry.Curves;
using Microsoft.Msagl.Core.Geometry;

namespace MsaglHelpers {
    public static class MsaglGeometryGraphHelpers {
        public static IEnumerable<PositionInfo<IDiagramItem>> GetGetNodesPositionInfo(GeometryGraph geometryGraph) {
            return geometryGraph.Nodes.Select(node => new PositionInfo<IDiagramItem>((IDiagramItem)node.UserData, Converter.MsaglPointToPointConvert(node.BoundingBox.LeftTop)));
        }
        public static PhyloTree CreatePhyloTrees(Graph<IDiagramItem> graph) {
            PhyloTree phyloTree = new PhyloTree();
            foreach(var node in graph.Nodes) {
                AddNode(phyloTree, node);
            }
            foreach(var edge in graph.Edges) {
                AddPhyloTreeEdge(phyloTree, edge.From, edge.To, edge.Weight);
            }
            return phyloTree;
        }
        public static GeometryGraph CreateGeometryGraph(Graph<IDiagramItem> graph) {
            GeometryGraph geomGraph = new GeometryGraph();
            foreach(var node in graph.Nodes) {
                AddNode(geomGraph, node);
            }
            foreach(var edge in graph.Edges) {
                AddEdge(geomGraph, edge.From, edge.To, edge.Weight);
            }
            return geomGraph;
        }
        public static Node AddNode(GeometryGraph geometryGraph, IDiagramItem item) {
            Node msaglNode = geometryGraph.FindNodeByUserData(item);
            if(msaglNode == null) {
                msaglNode = new Node(CreateCurve(item), item);
                geometryGraph.Nodes.Add(msaglNode);
            }
            return msaglNode;
        }
        public static void AddEdge(GeometryGraph geometryGraph, IDiagramItem parentNodeSource, IDiagramItem childNodeSource, double weight) {
            geometryGraph.Edges.Add(new Edge(AddNode(geometryGraph, parentNodeSource), AddNode(geometryGraph, childNodeSource)) { Weight = (int)weight });
        }
        public static void AddPhyloTreeEdge(PhyloTree phyloTree, IDiagramItem parentNodeSource, IDiagramItem childNodeSource, double weight = 1) {
            phyloTree.Edges.Add(new PhyloEdge(AddNode(phyloTree, parentNodeSource), AddNode(phyloTree, childNodeSource)) { Weight = (int)weight });
        }
        public static ICurve CreateCurve(IDiagramItem item) {
            return CurveFactory.CreateRectangle(item.ActualSize.Width, item.ActualSize.Height, new Point());
        }
    }
}
